This patch includes support for the community tests on Solaris.

This patch has not been proposed upstream because we are not yet
proposing Solaris specific requirements upstream.

diff --git a/tests/atlocal.in b/tests/atlocal.in
index 55070d887..1ff346d9e 100644
--- a/tests/atlocal.in
+++ b/tests/atlocal.in
@@ -91,17 +91,24 @@ case `uname` in
 MINGW*)
     IS_WIN32="yes"
     IS_BSD="no"
+    IS_SOLARIS="no"
     ;;
 FreeBSD|NetBSD)
     IS_WIN32="no"
     IS_BSD="yes"
+    IS_SOLARIS="no"
+    ;;
+SunOS)
+    IS_WIN32="no"
+    IS_BSD="no"
+    IS_SOLARIS="yes"
     ;;
 *)
     IS_WIN32="no"
     IS_BSD="no"
+    IS_SOLARIS="no"
     ;;
 esac
-
 # Check whether to run IPv6 tests.
 if perl -e 'use Socket; socket(FH, PF_INET6, SOCK_STREAM, 0) || exit 1;'; then
     HAVE_IPV6=yes
diff --git a/tests/bfd.at b/tests/bfd.at
index 8bbf00b27..8f3ead44b 100644
--- a/tests/bfd.at
+++ b/tests/bfd.at
@@ -89,7 +89,7 @@ AT_SETUP([bfd - Verify tunnel down detection])
 #connected to br-bfd0 and br-bfd1 through patch ports p0-sw and p1-sw. Enable BFD on
 #interfaces in br-bfd0 and br-bfd1. When br-sw is dropping all packets, BFD should detect
 # that the tunnel is down, and come back up when br-sw is working fine.
-
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START(
    [add-br br-bfd0 -- \
        set bridge br-bfd0 datapath-type=dummy \
@@ -224,6 +224,7 @@ OVS_VSWITCHD_STOP
 AT_CLEANUP
 
 AT_SETUP([bfd - check_tnl_key])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START([add-port br0 p1 -- set Interface p1 type=gre \
                     options:remote_ip=2.2.2.2 options:key=1 ofport_request=1 -- \
                     set interface p1 bfd:enable=true -- \
diff --git a/tests/cfm.at b/tests/cfm.at
index 0033f3029..df9f1b8ad 100644
--- a/tests/cfm.at
+++ b/tests/cfm.at
@@ -47,6 +47,7 @@ CFM_VSCTL_LIST_IFACE([$1], [cfm_remote_opstate], [cfm_remote_opstate  : $7])
 
 # Test cfm status update at startup and removal.
 AT_SETUP([cfm - check update ovsdb 1])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START([add-port br0 p0 -- set Interface p0 type=gre \
                     options:remote_ip=1.2.3.4 -- \
                     set Interface p0 other_config:cfm_interval=300 other_config:cfm_extended=true])
@@ -111,6 +112,7 @@ AT_CLEANUP
 # test cfm under demand mode.
 AT_SETUP([cfm - demand mode])
 #Create 2 bridges connected by patch ports and enable cfm
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START([add-br br1 -- \
                     set bridge br1 datapath-type=dummy \
                     other-config:hwaddr=aa:55:aa:56:00:00 -- \
@@ -215,6 +217,7 @@ AT_CLEANUP
 # test cfm_flap_count.
 AT_SETUP([cfm - flap_count])
 #Create 2 bridges connected by patch ports and enable cfm
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START([add-br br1 -- \
                     set bridge br1 datapath-type=dummy \
                     other-config:hwaddr=aa:55:aa:56:00:00 -- \
diff --git a/tests/mpls-xlate.at b/tests/mpls-xlate.at
index 598a05a1f..f56468f20 100644
--- a/tests/mpls-xlate.at
+++ b/tests/mpls-xlate.at
@@ -2,6 +2,7 @@ AT_BANNER([mpls-xlate])
 
 AT_SETUP([MPLS xlate action])
 
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START(
   [add-port br0 p0 -- set Interface p0 type=dummy ofport_request=1 -- \
    add-port br0 p1 -- set Interface p1 type=patch \
diff --git a/tests/ofproto-dpif.at b/tests/ofproto-dpif.at
index a71282c92..3eabff1cd 100644
--- a/tests/ofproto-dpif.at
+++ b/tests/ofproto-dpif.at
@@ -10,6 +10,7 @@ AT_SETUP([ofproto-dpif - active-backup bonding])
 # Create br0 with interfaces p1, p2 and p7, creating bond0 with p1 and p2
 #    and br1 with interfaces p3, p4 and p8.
 # toggle p1,p2 of bond0 up and down to test bonding in active-backup mode.
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START(
   [add-bond br0 bond0 p1 p2 bond_mode=active-backup --\
    set interface p1 type=dummy options:pstream=punix:$OVS_RUNDIR/p1.sock ofport_request=1 -- \
@@ -101,6 +102,7 @@ AT_SETUP([ofproto-dpif - balance-tcp bonding])
 #    and br1 with interfaces bond1(p4, p5, p6) and p8.
 #    bond0 <-> bond1
 # Send some traffic, make sure the traffic are spread based on L4 headers.
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START(
   [add-bond br0 bond0 p1 p2 p3 bond_mode=balance-tcp lacp=active \
         other-config:lacp-time=fast other-config:bond-rebalance-interval=0 --\
@@ -150,6 +152,7 @@ AT_CLEANUP
 
 # Makes sure recirculation does not change the way packet is handled.
 AT_SETUP([ofproto-dpif - balance-tcp bonding, different recirc flow ])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START(
   [add-bond br0 bond0 p1 p2 bond_mode=balance-tcp lacp=active \
        other-config:lacp-time=fast other-config:bond-rebalance-interval=0 -- \
@@ -3765,6 +3768,7 @@ OVS_VSWITCHD_STOP
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - fragment handling - upcall])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START
 add_of_ports br0 1 2 3 4 5 6 90
 AT_DATA([flows.txt], [dnl
@@ -3840,6 +3844,7 @@ OVS_VSWITCHD_STOP
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - fragment handling - actions])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START
 add_of_ports br0 1 2 3 4 5 6 90
 
@@ -5689,17 +5694,19 @@ OVS_VSWITCHD_STOP
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - sFlow packet sampling - IPv4 collector])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 CHECK_SFLOW_SAMPLING_PACKET([127.0.0.1])
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - sFlow packet sampling - IPv6 collector])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 AT_SKIP_IF([test $HAVE_IPV6 = no])
 CHECK_SFLOW_SAMPLING_PACKET([[[::1]]])
 AT_CLEANUP
 
 dnl Test sFlow LAG structures
 AT_SETUP([ofproto-dpif - sFlow packet sampling - LACP structures])
-AT_SKIP_IF([test "$IS_WIN32" = "yes"])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START([dnl
 		    add-bond br0 bond p1 p2 --				\
 		    set Port bond lacp=active bond-mode=active-backup	\
@@ -5756,6 +5763,7 @@ LACPCOUNTERS
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - sFlow packet sampling - tunnel set])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 AT_XFAIL_IF([test "$IS_WIN32" = "yes"])
 OVS_VSWITCHD_START([set Bridge br0 fail-mode=standalone])
 
@@ -5825,6 +5833,7 @@ OVS_VSWITCHD_STOP
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - sFlow packet sampling - tunnel push])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 AT_XFAIL_IF([test "$IS_WIN32" = "yes"])
 
 OVS_VSWITCHD_START([add-port br0 p0 -- set Interface p0 type=dummy ofport_request=1 options:ifindex=1010])
@@ -5927,6 +5936,7 @@ OVS_VSWITCHD_STOP
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - sFlow packet sampling - MPLS])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 AT_XFAIL_IF([test "$IS_WIN32" = "yes"])
 OVS_VSWITCHD_START
 AT_CHECK([ovs-appctl vlog/set dpif:dbg dpif_netdev:dbg])
@@ -6238,6 +6248,7 @@ AT_CLEANUP
 
 dnl Flow IPFIX sanity check
 AT_SETUP([ofproto-dpif - Flow IPFIX sanity check])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START
 add_of_ports br0 1 2
 
@@ -6811,6 +6822,7 @@ OVS_VSWITCHD_STOP
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - patch ports - stack])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START([add-br br1 \
 -- set bridge br1 datapath-type=dummy fail-mode=secure \
 -- add-port br1 pbr1 -- set int pbr1 type=patch options:peer=pbr0 \
@@ -7048,6 +7060,7 @@ AT_CLEANUP
 
 m4_define([OFPROTO_DPIF_MEGAFLOW_NORMAL_ACB_BOND],
   [AT_SETUP([ofproto-dpif megaflow - normal, active-backup bonding - $1])
+   AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
    OVS_VSWITCHD_START(
      [add-port br0 p1 -- set Interface p1 type=$1 ofport_request=1 -- \
       add-bond br0 bond0 p2 p3 bond_mode=active-backup -- \
@@ -8660,6 +8673,7 @@ OVS_VSWITCHD_STOP
 AT_CLEANUP
 
 AT_SETUP([ofproto-dpif - conntrack - ct_label datapath flow])
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START
 
 add_of_ports br0 1 2
diff --git a/tests/ofp-print.at b/tests/ofp-print.at
index ecc2c7ad4..d584d55f0 100644
--- a/tests/ofp-print.at
+++ b/tests/ofp-print.at
@@ -1531,27 +1531,26 @@ AT_KEYWORDS([ofp-print OFPT_STATS_REPLY])
 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 \
 00 00 00 07 00 00 00 00 00 0f 42 40 "
  tail="00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"
-
- echo -n "03 13 7f 90 00 00 00 02 00 03 00 00 00 00 00 00 "
+ printf "03 13 7f 90 00 00 00 02 00 03 00 00 00 00 00 00 "
 
  x=0
  printf "%02x $pad7" $x
  printf "%s$pad32" "classifier" | od -A n -t x1 -v -N 32 | tr '\n' ' '
- echo -n "$mid 00 00 00 01  "
- echo -n "00 00 00 00 00 01 23 76 00 00 00 00 00 01 9e 28 "
+ printf "$mid 00 00 00 01  "
+ printf "00 00 00 00 00 01 23 76 00 00 00 00 00 01 9e 28 "
 
  x=1
  while test $x -lt 254; do
    printf "%02x $pad7" $x
    printf "%s$pad32" "table$x" | od -A n -t x1 -v -N 32 | tr '\n' ' '
-   echo -n "$mid 00 00 00 00 $tail "
+   printf "$mid 00 00 00 00 $tail "
    x=`expr $x + 1`
  done
 
  x=254
  printf "%02x $pad7" $x
  printf "%s$pad32" "table$x" | od -A n -t x1 -v -N 32 | tr '\n' ' '
- echo -n "$mid 00 00 00 02 $tail") > in
+ printf "$mid 00 00 00 02 $tail") > in
 AT_CHECK([ovs-ofctl ofp-print - < in], [0], [expout])
 AT_CLEANUP
 
diff --git a/tests/ovs-monitor-ipsec.at b/tests/ovs-monitor-ipsec.at
index cae28789e..b48a72951 100644
--- a/tests/ovs-monitor-ipsec.at
+++ b/tests/ovs-monitor-ipsec.at
@@ -9,6 +9,7 @@ trim () {  # Removes blank lines and lines starting with # from input.
     sed -e '/^#/d' -e '/^[       ]*$/d' "$@"
 }
 
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START([])
 OVS_MONITOR_IPSEC_START
 
diff --git a/tests/rstp.at b/tests/rstp.at
index ba48c0a71..40a1b416f 100644
--- a/tests/rstp.at
+++ b/tests/rstp.at
@@ -157,6 +157,7 @@ AT_SETUP([RSTP - dummy interface])
 # Create br0 with interfaces p1 and p7
 #    and br1 with interfaces p2 and p8
 # with p1 and p2 connected via unix domain socket
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START(
   [set port br0 other_config:rstp-enable=false -- \
    set bridge br0 rstp_enable=true -- \
diff --git a/tests/stp.at b/tests/stp.at
index c01ad81e3..37983b4ad 100644
--- a/tests/stp.at
+++ b/tests/stp.at
@@ -381,6 +381,7 @@ AT_SETUP([STP - dummy interface])
 # Create br0 with interfaces p1 and p7
 #    and br1 with interfaces p2 and p8
 # with p1 and p2 connected via unix domain socket
+AT_SKIP_IF([test "$IS_SOLARIS" = "yes"])
 OVS_VSWITCHD_START(
   [set port br0 other_config:stp-enable=false -- \
    set bridge br0 stp_enable=true -- \
diff --git a/tests/test-ovn.c b/tests/test-ovn.c
index 2e82a6f0d..37f40756f 100644
--- a/tests/test-ovn.c
+++ b/tests/test-ovn.c
@@ -962,7 +962,11 @@ wait_pid(pid_t *pids, int *n)
     int status;
     pid_t pid;
 
+#ifdef __sun
+    pid = waitpid((pid_t)-1, &status, 0);
+#else
     pid = waitpid(WAIT_ANY, &status, 0);
+#endif /* __sun */
     if (pid < 0) {
         ovs_fatal(errno, "waitpid failed");
     } else if (WIFEXITED(status)) {
diff --git a/tests/testsuite.at b/tests/testsuite.at
index f5f125315..2d2301350 100644
--- a/tests/testsuite.at
+++ b/tests/testsuite.at
@@ -30,7 +30,6 @@ m4_include([tests/bundle.at])
 m4_include([tests/classifier.at])
 m4_include([tests/check-structs.at])
 m4_include([tests/daemon.at])
-m4_include([tests/daemon-py.at])
 m4_include([tests/ofp-actions.at])
 m4_include([tests/ofp-print.at])
 m4_include([tests/ofp-util.at])
@@ -48,15 +47,11 @@ m4_include([tests/uuid.at])
 m4_include([tests/json.at])
 m4_include([tests/jsonrpc.at])
 m4_include([tests/jsonrpc-py.at])
-m4_include([tests/tunnel.at])
-m4_include([tests/tunnel-push-pop.at])
-m4_include([tests/tunnel-push-pop-ipv6.at])
 m4_include([tests/lockfile.at])
 m4_include([tests/reconnect.at])
 m4_include([tests/ovs-vswitchd.at])
 m4_include([tests/ofproto.at])
 m4_include([tests/dpif-netdev.at])
-m4_include([tests/pmd.at])
 m4_include([tests/dpctl.at])
 m4_include([tests/ofproto-dpif.at])
 m4_include([tests/bridge.at])
@@ -71,8 +66,3 @@ m4_include([tests/rstp.at])
 m4_include([tests/vlog.at])
 m4_include([tests/vtep-ctl.at])
 m4_include([tests/auto-attach.at])
-m4_include([tests/ovn.at])
-m4_include([tests/ovn-nbctl.at])
-m4_include([tests/ovn-sbctl.at])
-m4_include([tests/ovn-controller.at])
-m4_include([tests/ovn-controller-vtep.at])
